### ISSUE TITLE
Python scripts

### ISSUE THREAD

**Issue by q0p1us** [AlexeyAB/darknet#621]
Hi,
Thanks for adding many improvements to this repo over the original one.
The original repo has a very useful python script (https://github.com/pjreddie/darknet/tree/master/python) for calling the detection function. I tried that script on this repo but it does not work (errors like `darknet.so: undefined symbol: load_network`).
Could you please add a similar script to your repo?
Thank you very much!

**Comment by xingyewuyu** [comment]
The include path for the so lib is not right.

**Comment by AlexeyAB** [comment]
I added support of Python 2.x: https://www.python.org/downloads/release/python-2714/

* **on Linux:** run `darknet/darknet.py` (use `LIBSO=1` in the `Makefile` and compile)

* **on Windows:** run `darknet/build/darknet/x64/darknet.py` (compile `darknet/build/darknet/yolo_cpp_dll.sln`)

I tested it on Windows only:

![image](https://user-images.githubusercontent.com/4096485/38699351-726e67bc-3ea0-11e8-8e69-7371e10d7d1b.png)

**Comment by Jacky3213** [comment]
@AlexeyAB  Sir, thanks for your contribution. Here is my question. I get the result as you show, but i find it is not right when draw the bouding boxes on the image.  The  width and height of image(dog.jpg) are 768 and 576. I am really confused with it. Thirty for your reply.

**Comment by q0p1us** [comment]
@AlexeyAB Thanks a lot! It seems to still have bugs, though. I'm on Ubuntu 16.04.

With this build setting:
```
GPU=1
CUDNN=0
CUDNN_HALF=0
OPENCV=1
DEBUG=0
OPENMP=1
LIBSO=1
```
when running `darknet.py` it successfully loaded the weights but then segmentation fault:
```
106 detection
Loading weights from yolov3.weights...
 seen 64 
Done!
Loaded - names_list: data/coco.names, classes = 80 
Segmentation fault (core dumped)
```
Setting `GPU=0` I obtained the following error when running `darknet.py`:
`AttributeError: ./darknet.so: undefined symbol: cuda_set_device`

My GPU only have 2GB of memory so that may be the cause of the "segmentation fault". Could you please fix the bug when setting `GPU=0`? 
Thank you very much in advance!

**Comment by q0p1us** [comment]
I have tried running on a K80 GPU with 11GB of memory and it also resulted in `segmentation fault (core dumped)`.

**Comment by AlexeyAB** [comment]
Try to use the last version of code and these params in the Makefile:
```
GPU=1
CUDNN=1
CUDNN_HALF=0
OPENCV=1
AVX=0
OPENMP=0
LIBSO=1
DEBUG=0
```

**Comment by saihv** [comment]
@AlexeyAB Thanks for adding Python support! Just a quick question, for the same image, same cfg and similar Makefile parameters, the original version of Darknet reports 14 ms for a detection through the Python `detect()` function, whereas this version reports time taken of about 46 ms. Would you have any thoughts on where the bottleneck could be? My test was using Tiny YOLO (custom trained) on a GTX 1080 Ti.

For reference, my compilation flags are:
`GPU=1
CUDNN=1
CUDNN_HALF=0
OPENCV=1
AVX=0
OPENMP=1
LIBSO=1`

My way of timing was:

```
    start = time.time()
    r = detect(net, meta, "test.jpg", 0.25)
    print time.time()-start
```

**Comment by saihv** [comment]
@AlexeyAB Thanks for adding Python support! Just a quick question, for the same image, same cfg and similar Makefile parameters, the original version of Darknet reports 19 ms for a detection through the Python `detect()` function, whereas this version reports time taken of about 46 ms. Would you have any thoughts on where the bottleneck could be? My test was using Tiny YOLO (custom trained) on a GTX 1080 Ti.

For reference, my compilation flags are:
`GPU=1
CUDNN=1
CUDNN_HALF=0
OPENCV=1
AVX=0
OPENMP=1
LIBSO=1`

My way of timing was:

```
    start = time.time()
    r = detect(net, meta, "test.jpg", 0.25)
    print time.time()-start
```

Just for reference, in your version, the `./darknet detector test` command reports that the same prediction was done in 5.4 ms.

**Comment by AlexeyAB** [comment]
@saihv Hi,

If you try to check time only for these lines, can you get the same prediction time?
* Original: https://github.com/pjreddie/darknet/blob/508381b37fe75e0e1a01bcb2941cb0b31eb0e4c9/python/darknet.py#L127-L142
* This repo: https://github.com/AlexeyAB/darknet/blob/5e3dcb6f34868e341466b57b13ad63f86b337250/darknet.py#L128-L144

I.e. measure time without image loading.

**Comment by saihv** [comment]
Just checked that: it's actually taking a long time during `predict_image()`, but in the image pre-processing stage. Specifically, in `image.c`, the resize function is taking 26 ms and the complete letterboxing operation is taking 35 ms, with ~40ms for the Python `predict_image()` function, as opposed to ~20ms in the original repo . 

It's strange because even the C++ binary will have to go through the same functions. Do you see something similar in your tests?

**Comment by AlexeyAB** [comment]
@saihv Thanks!


Try to change this code: https://github.com/AlexeyAB/darknet/blob/5e3dcb6f34868e341466b57b13ad63f86b337250/src/image.c#L1644-L1668

To this and re-compile:
```cpp
static float get_pixel(image m, int x, int y, int c)
{
    assert(x < m.w && y < m.h && c < m.c);
    return m.data[c*m.h*m.w + y*m.w + x];
}
static float get_pixel_extend(image m, int x, int y, int c)
{
    if(x < 0 || x >= m.w || y < 0 || y >= m.h) return 0;
    /*
    if(x < 0) x = 0;
    if(x >= m.w) x = m.w-1;
    if(y < 0) y = 0;
    if(y >= m.h) y = m.h-1;
    */
    if(c < 0 || c >= m.c) return 0;
    return get_pixel(m, x, y, c);
}
static void set_pixel(image m, int x, int y, int c, float val)
{
    if (x < 0 || y < 0 || c < 0 || x >= m.w || y >= m.h || c >= m.c) return;
    assert(x < m.w && y < m.h && c < m.c);
    m.data[c*m.h*m.w + y*m.w + x] = val;
}
static void add_pixel(image m, int x, int y, int c, float val)
{
    assert(x < m.w && y < m.h && c < m.c);
    m.data[c*m.h*m.w + y*m.w + x] += val;
}
```

**Comment by saihv** [comment]
That did the trick! Now resizing takes 13ms, and the total prediction takes 20ms. (also had to change the declarations in `image.h` to match `static`)

**Comment by AlexeyAB** [comment]
@saihv I added this fix to the repository.

**Comment by JayMan69** [comment]
I am having the same issue as user q0p1us.
Ubuntu version 16.04
Make file values
GPU=1
CUDNN=1
CUDNN_HALF=0
OPENCV=1
AVX=0
OPENMP=0
LIBSO=1
DEBUG=0

I get a Segmentation Fault (core dumped) when running the command python darknet.py. However the running the ./darknet command itself has no issues whatsoever. I did a git pull of the github codebase as of 5/1.
Thank you in advance.

**Comment by AlexeyAB** [comment]
@Jacky3213 

* Did you download `yolov3.weights`? https://pjreddie.com/media/files/yolov3.weights
* Do you have all these files by these paths? 
https://github.com/AlexeyAB/darknet/blob/a0e288f0dca328b8b12592cc19da9bb7c29646f4/darknet.py#L153-L155

* What GPU do you use?
* What Python version do you use?

**Comment by AlexeyAB** [comment]
@JayMan69 
It seems this issue is for python3.6.3, but it can work well for python <= 3.5.2: https://github.com/AlexeyAB/darknet/issues/732#issuecomment-386213024

**Comment by AlexeyAB** [comment]
@JayMan69 Can this issue be solved using new Python script on python3.6.3?

https://github.com/AlexeyAB/darknet/blob/master/darknet.py

**Comment by AlexeyAB** [comment]
@q0p1us @xingyewuyu @Jacky3213 @saihv @JayMan69 
There is added new Python script that can be used with Python 2.x and 3x: https://github.com/AlexeyAB/darknet/pull/754

**Comment by JayMan69** [comment]
@AlexeyAB,
Yes that worked. There was a small fix on line 121 of darknet.py where i had to replace libdarknet.so with ./darknet.so.

Thank you so much!

**Comment by FilipaRamos** [comment]
I keep on getting ```Fatal Python error: PyThreadState_Get: no current thread``` using either libdarknet.so or darknet.so. Does anybody have any idea what this could be about?

**Comment by AlexeyAB** [comment]
@FilipaRamos 
Do you use new Python-script? https://github.com/AlexeyAB/darknet/blob/cda8171feb76bcb405350fd8341d42a0300e2f4b/darknet.py

Or old Python-script? https://github.com/AlexeyAB/darknet/blob/89354d0a0ce6fbb22ff262658045cdb8796ff6fd/darknet.py

And what date of your Darknet-code? Try to update your code from GitHub.

**Comment by FilipaRamos** [comment]
I am using the newest script with python 3.5.2. I also pulled the latest Darknet code from this repo. The error still happens :\

**Comment by FilipaRamos** [comment]
@AlexeyAB should have said it before but unfortunately I'm on windows (and using anaconda)

**Comment by FilipaRamos** [comment]
@AlexeyAB I am able to run the script when I am on the base python version. For any other, it gives that error. Anyway, it's not an issue with the script itself but with python configuration. Thanks for all the help.

**Comment by FilipaRamos** [comment]
Is it possible to run detections on a custom trained dataset? I see no detections for now on custom datasets.

**Comment by AlexeyAB** [comment]
@FilipaRamos Yes, what command do you use? And how many iterations did you train?

----

You should use such as: https://github.com/AlexeyAB/darknet#custom-object-detection
`darknet.exe detector test data/obj.data yolo-obj.cfg yolo-obj_8000.weights`

----

Or for Python change this line: https://github.com/AlexeyAB/darknet/blob/3ebcc647b651a4a3c717eff2a3087127e5707e0c/build/darknet/x64/darknet.py#L399
to this:
`print(performDetect("data/dog.jpg", 0.25, "yolo-obj.cfg", "yolo-obj.cfg", "./data/obj.data", True,  False,  False))`

**Comment by FilipaRamos** [comment]
I have trained the model on the C++ darknet and all detections are fine. However, when I try to run detections with my weight and cfg files on the python wrapper, it outputs no detections. I am using the command you wrote above ^.

**Comment by AlexeyAB** [comment]
@FilipaRamos I fixed the above my typo, use this code, instead of `image5.jpg` use your image filename:
```
if __name__ == "__main__":
    print(performDetect("image5.jpg", 0.25, "yolo-obj.cfg", "yolo-obj_8000.weights", "./data/obj.data", True,  False,  False))
```

![image](https://user-images.githubusercontent.com/4096485/39995518-1de045e6-5785-11e8-9fe5-e52fa9d28667.png)

**Comment by FilipaRamos** [comment]
I am not sure what is happening, however, I can not get any detections. I am fairly certain that I am using the right command. I tried once more on the C++ and it works perfectly.

```print(performDetect("./data/obj/0005_0000000010.jpg", 0.25, "yolov3.cfg", "backup/yolov3_12100.weights", "./data/obj.data", True, False, False))```

**Comment by FilipaRamos** [comment]
I am not familiar with the code however, the for cycle on the function detect() seems to be skipped entirely. Is this normal behaviour when there are detections on the picture?

**Comment by AlexeyAB** [comment]
@FilipaRamos What command do you use to successfully detect using darknet.exe?

**Comment by FilipaRamos** [comment]
Using darknet.exe I do: ```darknet detector test data/obj.data yolov3.cfg backup/yolov3_12100.weights data/obj/0005_0000000010.jpg```

**Comment by FilipaRamos** [comment]
@AlexeyAB I always get this print at the beggining of the run:

```Environment variables indicated a CPU run, but we didn't find ´yolo_cpp_dll_nogpu.dll`. Trying a GPU run anyway.```` 

Could this be related to me getting no detections?

**Comment by AlexeyAB** [comment]
@FilipaRamos I just fixed it (DLL/SO-library and darknet.py). 
Update your code from GitHub, **rebuild DLL-library** and run new darknet.py

Earlier darknet.py required batch=1 in cfg-file, but now it doesn't require change batch in cfg-file.

**Comment by FilipaRamos** [comment]
@AlexeyAB Just tested it and its working great! Thanks a lot for all your work :)

**Comment by wting861006** [comment]
@AlexeyAB Hi,I have two gpus.I want to let darknet_video.py run at gpu0 with video0.avi,and gpu1 with video1.avi.How to set gpu_id?
